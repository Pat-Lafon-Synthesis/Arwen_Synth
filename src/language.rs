use std::collections::HashMap;
use std::fmt::{Debug, Display};
use std::hash::Hash;
use std::rc::Rc;

use itertools::Itertools;

use crate::types::{BaseType, RefinementType, Signature};

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Tree<T: Display> {
    Leaf,
    Node(T, Box<Tree<T>>, Box<Tree<T>>),
}

impl<T: Display> Display for Tree<T> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{}",
            match self {
                Tree::Leaf => "leaf".to_string(),
                Tree::Node(n, t1, t2) => format!("node {} ({}) ({})", n, t1, t2),
            }
        )
    }
}

#[derive(Debug, Clone, PartialEq, Hash, Eq, PartialOrd, Ord)]
pub enum Constant {
    Int(i64),
    Bool(bool),
    IntList(Vec<i64>),
    IntTree(Tree<i64>),
}

impl Display for Constant {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{}",
            match self {
                Constant::Int(i) => i.to_string(),
                Constant::Bool(b) => b.to_string(),
                Constant::IntList(_) => "list".to_string(),
                Constant::IntTree(_) => "tree".to_string(),
            }
        )
    }
}

#[derive(Clone)]
pub struct Operation {
    pub name: String,
    pub sig: Signature<BaseType>,
    pub code: Rc<dyn Fn(&Vec<Constant>) -> Result<Constant, InvalidProg>>,
}

impl Operation {
    pub fn execute(&self, args: &Vec<Constant>) -> Result<Constant, InvalidProg> {
        (self.code)(args)
    }
}

impl Display for Operation {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.name)
    }
}

impl std::fmt::Debug for Operation {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Operation")
            .field("name", &self.name)
            .field("sig", &self.sig)
            .finish()
    }
}

impl Hash for Operation {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.name.hash(state);
    }
}

impl PartialEq for Operation {
    fn eq(&self, other: &Self) -> bool {
        self.name == other.name
    }
}

impl Eq for Operation {}

impl PartialOrd for Operation {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        self.name.partial_cmp(&other.name)
    }
}

impl Ord for Operation {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.name.cmp(&other.name)
    }
}

#[derive(Debug, Clone, Hash, PartialEq, Eq, PartialOrd, Ord)]
pub struct Variable {
    pub name: String,
    pub ty: RefinementType,
}

impl Display for Variable {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.name)
    }
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum ProgramNode {
    Constant(Constant),
    Variable(Variable, BaseType),
    Operation(Operation),
    If,
}

impl Display for ProgramNode {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ProgramNode::Constant(c) => write!(f, "{c}"),
            ProgramNode::Variable(v, _) => write!(f, "{v}"),
            ProgramNode::Operation(o) => write!(f, "{}", o),
            ProgramNode::If => write!(f, "if"),
        }
    }
}

type Environment = HashMap<Variable, Constant>;

#[derive(Debug, Clone, PartialEq, PartialOrd, Eq, Ord)]
pub struct TestCase {
    pub inputs: Vec<Constant>,
    pub output: Constant,
}

impl TestCase {
    fn into_env(&self) -> Environment {
        let TestCase { inputs, .. } = self;
        inputs
            .iter()
            .enumerate()
            .map(|(i, c)| {
                (
                    Variable {
                        name: format!("arg{i:?}"),
                        ty: Into::<BaseType>::into(c).into(),
                    },
                    c.clone(),
                )
            })
            .collect()
    }
}

impl Display for TestCase {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "({}) -> {}",
            self.inputs.iter().map(ToString::to_string).join(","),
            self.output
        )
    }
}

pub struct InvalidProg {}

#[derive(Debug, Clone, PartialEq, PartialOrd, Eq, Ord)]
pub struct Program {
    pub node: ProgramNode,
    pub args: Vec<Program>,
}

impl Display for Program {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match &self.node {
            ProgramNode::Constant(c) => write!(f, "{c}"),
            ProgramNode::Variable(v, _) => write!(f, "{v}"),
            ProgramNode::Operation(o) => write!(
                f,
                "{}({})",
                o,
                self.args.iter().map(ToString::to_string).join(",")
            ),
            ProgramNode::If => write!(
                f,
                "if ({}) {{{}}} else {{{}}}",
                self.args.get(0).unwrap(),
                self.args.get(1).unwrap(),
                self.args.get(2).unwrap()
            ),
        }
    }
}

impl Program {
    pub fn interpret(&self, e: &Environment) -> Result<Constant, InvalidProg> {
        let Program { node, args } = self;
        Ok(match node {
            ProgramNode::Constant(c) => c.clone(),
            ProgramNode::Variable(v, _) => e.get(v).unwrap().clone(),
            ProgramNode::Operation(o) => {
                let args = args.iter().map(|a| a.interpret(e)).try_collect()?;
                o.execute(&args)?
            }
            ProgramNode::If => {
                if args.get(0).unwrap().interpret(e)? == Constant::Bool(true) {
                    args.get(1).unwrap().interpret(e)?
                } else {
                    args.get(2).unwrap().interpret(e)?
                }
            }
        })
    }

    pub fn get_behavior(&self, tests: &Vec<TestCase>) -> Vec<TestCase> {
        tests
            .iter()
            .filter_map(|t| {
                self.interpret(&t.into_env()).ok().map(|output| TestCase {
                    inputs: t.inputs.clone(),
                    output,
                })
            })
            .collect()
    }

    pub fn passes_test_case(&self, t: &TestCase) -> bool {
        self.interpret(&t.into_env())
            .map_or(false, |output| output == t.output)
    }

    pub fn passes_all_test_cases(&self, v: &Vec<TestCase>) -> bool {
        v.iter().all(|t| self.passes_test_case(t))
    }
}
