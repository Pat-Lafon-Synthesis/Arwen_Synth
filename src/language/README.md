# Language Constructs

`InOutExamples.rs` contains code related to mapping input values to output values. Used in describing the behavior of components, providing input to output mappings from users, or the inductive specification of holes.

`language.rs` contains code related to program structure. Either complete programs/constant values, or in sketches containing holes.
