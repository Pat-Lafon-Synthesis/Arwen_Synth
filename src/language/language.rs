use std::collections::HashMap;
use std::fmt::{Debug, Display};
use std::hash::Hash;
use std::rc::Rc;

use ecta_rs::{ECTANode, Edge};
use itertools::Itertools;

use crate::data_structures::SynthCostFunc;
use crate::types::{Signature, TypeSystemBounds};
use crate::{SynthEcta, SynthEctaEdge};

use super::TestCase;

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Tree<T: Display> {
    Leaf,
    Node(T, Box<Tree<T>>, Box<Tree<T>>),
}

impl<T: Display> Display for Tree<T> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{}",
            match self {
                Tree::Leaf => "leaf".to_string(),
                Tree::Node(n, t1, t2) => format!("node {} ({}) ({})", n, t1, t2),
            }
        )
    }
}

#[derive(Debug, Clone, PartialEq, Hash, Eq, PartialOrd, Ord)]
pub enum Constant {
    Int(i64),
    Bool(bool),
    IntList(Vec<i64>),
    IntTree(Tree<i64>),
}

impl Display for Constant {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{}",
            match self {
                Constant::Int(i) => i.to_string(),
                Constant::Bool(b) => b.to_string(),
                Constant::IntList(l) => format!("[{}]", l.iter().rev().join(", ")),
                Constant::IntTree(_) => "tree".to_string(),
            }
        )
    }
}

impl From<i64> for Constant {
    fn from(i: i64) -> Self {
        Constant::Int(i)
    }
}

impl TryFrom<Constant> for i64 {
    type Error = InvalidProg;

    fn try_from(value: Constant) -> Result<Self, Self::Error> {
        match value {
            Constant::Int(i) => Ok(i),
            _ => Err(InvalidProg {}),
        }
    }
}

impl TryFrom<&Constant> for i64 {
    type Error = InvalidProg;

    fn try_from(value: &Constant) -> Result<Self, Self::Error> {
        match value {
            Constant::Int(i) => Ok(*i),
            _ => Err(InvalidProg {}),
        }
    }
}

impl From<bool> for Constant {
    fn from(b: bool) -> Self {
        Constant::Bool(b)
    }
}

impl TryFrom<Constant> for bool {
    type Error = InvalidProg;

    fn try_from(value: Constant) -> Result<Self, Self::Error> {
        match value {
            Constant::Bool(b) => Ok(b),
            _ => Err(InvalidProg {}),
        }
    }
}

impl TryFrom<&Constant> for bool {
    type Error = InvalidProg;

    fn try_from(value: &Constant) -> Result<Self, Self::Error> {
        match value {
            Constant::Bool(b) => Ok(*b),
            _ => Err(InvalidProg {}),
        }
    }
}

impl From<Vec<i64>> for Constant {
    fn from(v: Vec<i64>) -> Self {
        Constant::IntList(v)
    }
}

impl TryFrom<Constant> for Vec<i64> {
    type Error = InvalidProg;

    fn try_from(value: Constant) -> Result<Self, Self::Error> {
        match value {
            Constant::IntList(v) => Ok(v),
            _ => Err(InvalidProg {}),
        }
    }
}

impl From<Tree<i64>> for Constant {
    fn from(t: Tree<i64>) -> Self {
        Constant::IntTree(t)
    }
}

impl TryFrom<Constant> for Tree<i64> {
    type Error = InvalidProg;

    fn try_from(value: Constant) -> Result<Self, Self::Error> {
        match value {
            Constant::IntTree(t) => Ok(t),
            _ => Err(InvalidProg {}),
        }
    }
}

pub type SynthFuncType = dyn Fn(&Vec<Constant>) -> Result<Constant, InvalidProg>;

#[derive(Clone)]
pub struct Operation<T> {
    pub name: String,
    pub sig: Signature<T>,
    pub code: Rc<SynthFuncType>,
}

impl<T> Operation<T> {
    pub fn execute(&self, args: &Vec<Constant>) -> Result<Constant, InvalidProg> {
        (self.code)(args)
    }
}

impl<T> Display for Operation<T> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.name)
    }
}

impl<T: Debug> std::fmt::Debug for Operation<T> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Operation")
            .field("name", &self.name)
            .field("sig", &self.sig)
            .finish()
    }
}

impl<T> Hash for Operation<T> {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.name.hash(state);
    }
}

impl<T> PartialEq for Operation<T> {
    fn eq(&self, other: &Self) -> bool {
        self.name == other.name
    }
}

impl<T> Eq for Operation<T> {}

impl<T> PartialOrd for Operation<T> {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        self.name.partial_cmp(&other.name)
    }
}

impl<T> Ord for Operation<T> {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.name.cmp(&other.name)
    }
}

#[derive(Debug, Clone, Hash, PartialEq, Eq, PartialOrd, Ord)]
pub struct Variable<T: TypeSystemBounds> {
    pub name: String,
    pub ty: T,
}

impl<T: TypeSystemBounds> Display for Variable<T> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.name)
    }
}

// Program node that can be used to create a block/trace
// Basically no if statements or recursive calls
#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub enum LinearProgramNode<T: TypeSystemBounds> {
    Constant(Constant),
    Variable(Variable<T>),
    Operation(Operation<T>),
}

impl<T: TypeSystemBounds> LinearProgramNode<T> {
    /// Produces a sorted list of next nodes for a given edge
    pub fn ecta_next_program_nodes(
        &self,
        ecta: &SynthEcta<T>,
        start_node: ECTANode,
    ) -> Option<Vec<ECTANode>> {
        self.ecta_next_edges(ecta, start_node).map(|edges| {
            edges
                .into_iter()
                .sorted_by_key(|e| e.edge_num)
                // Skipping the edge that is the return type
                .skip(1)
                .map(|e| e.nodeidx)
                .collect_vec()
        })
    }

    fn ecta_next_edges(
        &self,
        ecta: &SynthEcta<T>,
        start_node: ECTANode,
    ) -> Option<Vec<Edge<SynthEctaEdge<T>, ()>>> {
        // Get the group of edges for this operation
        let g_group = ecta
            .get_edges(start_node)
            .sorted_by_key(|e| e.data.clone())
            .group_by(|e| e.data.clone());

        let g = g_group.into_iter();

        let mut edge_nodes = g
            .filter(|e| matches!(&e.0, SynthEctaEdge::P(o2) if self == o2))
            .collect_vec();

        if edge_nodes.is_empty() {
            return None;
        }

        // Check that there was only one option
        assert!(edge_nodes.len() == 1);

        // All the edges for the operation in the ecta
        let x = Some(edge_nodes.pop().unwrap().1.cloned().collect_vec());
        x
    }
}

impl<T: TypeSystemBounds> Display for LinearProgramNode<T> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            LinearProgramNode::Constant(c) => write!(f, "{c}"),
            LinearProgramNode::Variable(v) => write!(f, "{v}"),
            LinearProgramNode::Operation(o) => write!(f, "{}", o),
        }
    }
}

impl<T: TypeSystemBounds> TryFrom<LinearProgramNode<T>> for Operation<T> {
    type Error = ();

    fn try_from(value: LinearProgramNode<T>) -> Result<Self, Self::Error> {
        match value {
            LinearProgramNode::Constant(_) => Err(()),
            LinearProgramNode::Variable(_) => Err(()),
            LinearProgramNode::Operation(o) => Ok(o),
        }
    }
}

impl<T: TypeSystemBounds> TryFrom<LinearProgramNode<T>> for Variable<T> {
    type Error = ();

    fn try_from(value: LinearProgramNode<T>) -> Result<Self, Self::Error> {
        match value {
            LinearProgramNode::Constant(_) => Err(()),
            LinearProgramNode::Variable(v) => Ok(v),
            LinearProgramNode::Operation(_) => Err(()),
        }
    }
}

impl<T: TypeSystemBounds> TryFrom<LinearProgramNode<T>> for Constant {
    type Error = ();

    fn try_from(value: LinearProgramNode<T>) -> Result<Self, Self::Error> {
        match value {
            LinearProgramNode::Constant(c) => Ok(c),
            LinearProgramNode::Variable(_) => Err(()),
            LinearProgramNode::Operation(_) => Err(()),
        }
    }
}

impl<T: TypeSystemBounds> TryFrom<ProgramNode<T>> for LinearProgramNode<T> {
    type Error = ();

    fn try_from(value: ProgramNode<T>) -> Result<Self, ()> {
        match value {
            ProgramNode::Constant(c) => Ok(LinearProgramNode::Constant(c)),
            ProgramNode::Variable(v) => Ok(LinearProgramNode::Variable(v)),
            ProgramNode::Operation(o) => Ok(LinearProgramNode::Operation(o)),
            ProgramNode::If => Err(()),
            ProgramNode::Rec(_) => Err(()),
        }
    }
}

impl<T: TypeSystemBounds> From<LinearProgramNode<T>> for ProgramNode<T> {
    fn from(value: LinearProgramNode<T>) -> Self {
        match value {
            LinearProgramNode::Constant(c) => ProgramNode::Constant(c),
            LinearProgramNode::Variable(v) => ProgramNode::Variable(v),
            LinearProgramNode::Operation(o) => ProgramNode::Operation(o),
        }
    }
}

#[derive(Debug, Clone)]
pub struct LinearProgram<T: TypeSystemBounds> {
    pub node: LinearProgramNode<T>,
    pub args: Vec<LinearProgram<T>>,
}

impl<T: TypeSystemBounds> LinearProgram<T> {
    pub fn new(node: LinearProgramNode<T>, args: Vec<LinearProgram<T>>) -> Self {
        Self { node, args }
    }

    pub fn is_trace_consistent(&self, ecta: &SynthEcta<T>, start_node: ECTANode) -> bool {
        if let Some(nodes) = self.node.ecta_next_program_nodes(ecta, start_node) {
            assert!(nodes.len() == self.args.len());
            self.args
                .iter()
                .zip(nodes.iter())
                .all(|(a, n)| a.is_trace_consistent(ecta, *n))
        } else {
            false
        }
    }
    pub fn contains_variable(&self) -> bool {
        match self.node {
            LinearProgramNode::Variable(_) => true,
            LinearProgramNode::Constant(_) => false,
            LinearProgramNode::Operation(_) => self.args.iter().any(|a| a.contains_variable()),
        }
    }

    pub fn interpret(&self, e: &Environment<T>) -> Result<Constant, InvalidProg> {
        let LinearProgram { node, args } = self;
        Ok(match node {
            LinearProgramNode::Constant(c) => c.clone(),
            LinearProgramNode::Variable(v) => e.get(v).unwrap().clone(),
            LinearProgramNode::Operation(o) => {
                let args = args.iter().map(|a| a.interpret(e)).try_collect()?;
                o.execute(&args)?
            }
        })
    }

    pub fn get_behavior(&self, tests: &[TestCase]) -> Vec<TestCase> {
        tests
            .iter()
            .filter_map(|t| {
                self.interpret(&t.into()).ok().map(|output| TestCase {
                    inputs: t.inputs.clone(),
                    output,
                })
            })
            .collect()
    }

    pub fn get_type(&self) -> T {
        match &self.node {
            LinearProgramNode::Constant(c) => c.clone().into(),
            LinearProgramNode::Variable(v) => v.ty.clone(),
            LinearProgramNode::Operation(o) => o.sig.output.clone(),
        }
    }
}

impl<T: TypeSystemBounds> Display for LinearProgram<T> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", Program::from(self.clone()))
    }
}

impl<T: TypeSystemBounds> SynthCostFunc for LinearProgram<T> {
    fn cost(&self) -> usize {
        match &self.node {
            LinearProgramNode::Variable(_) => 1,
            LinearProgramNode::Constant(_) => 2,
            LinearProgramNode::Operation(_) => {
                3 + self.args.iter().map(|a| a.cost()).sum::<usize>()
            }
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub enum ProgramNode<T: TypeSystemBounds> {
    Constant(Constant),
    Variable(Variable<T>),
    Operation(Operation<T>),
    If,
    Rec(T), // Recursive Call
}

impl<T: TypeSystemBounds> Display for ProgramNode<T> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ProgramNode::Constant(c) => write!(f, "{c}"),
            ProgramNode::Variable(v) => write!(f, "{v}"),
            ProgramNode::Operation(o) => write!(f, "{}", o),
            ProgramNode::If => write!(f, "if"),
            ProgramNode::Rec(t) => write!(f, "rec<{t}>"),
        }
    }
}

pub struct Environment<T: TypeSystemBounds>(HashMap<Variable<T>, Constant>);

impl<T: TypeSystemBounds> Environment<T> {
    pub fn new() -> Self {
        Environment(HashMap::new())
    }

    pub fn get(&self, v: &Variable<T>) -> Option<&Constant> {
        self.0.get(v)
    }
}

impl<T: TypeSystemBounds> Default for Environment<T> {
    fn default() -> Self {
        Self::new()
    }
}

impl<T: TypeSystemBounds> From<&TestCase> for Environment<T> {
    fn from(TestCase { inputs, .. }: &TestCase) -> Self {
        Environment(
            inputs
                .iter()
                .enumerate()
                .map(|(i, c)| {
                    (
                        Variable {
                            name: format!("arg{i:?}"),
                            ty: Into::<T>::into(c.clone()),
                        },
                        c.clone(),
                    )
                })
                .collect(),
        )
    }
}

impl<T: TypeSystemBounds> From<&Vec<Constant>> for Environment<T> {
    fn from(inputs: &Vec<Constant>) -> Self {
        Environment(
            inputs
                .iter()
                .enumerate()
                .map(|(i, c)| {
                    (
                        Variable {
                            name: format!("arg{i:?}"),
                            ty: Into::<T>::into(c.clone()),
                        },
                        c.clone(),
                    )
                })
                .collect(),
        )
    }
}

impl<T: TypeSystemBounds> From<Vec<Constant>> for Environment<T> {
    fn from(inputs: Vec<Constant>) -> Self {
        Environment(
            inputs
                .into_iter()
                .enumerate()
                .map(|(i, c)| {
                    (
                        Variable {
                            name: format!("arg{i:?}"),
                            ty: Into::<T>::into(c.clone()),
                        },
                        c,
                    )
                })
                .collect(),
        )
    }
}

#[derive(Debug)]
pub struct InvalidProg {}

#[derive(Debug, Clone /* , PartialEq, Eq */)]
pub struct Program<T: TypeSystemBounds> {
    pub node: ProgramNode<T>,
    pub args: Vec<Program<T>>,
}

impl<T: TypeSystemBounds> Program<T> {
    pub fn new(node: ProgramNode<T>, args: Vec<Program<T>>) -> Self {
        Program { node, args }
    }

    pub fn interpret(
        &self,
        e: &Environment<T>,
        full_program: &Self,
    ) -> Result<Constant, InvalidProg> {
        let Program { node, args } = self;
        Ok(match node {
            ProgramNode::Constant(c) => c.clone(),
            ProgramNode::Variable(v) => e.get(v).unwrap().clone(),
            ProgramNode::Operation(o) => {
                let args = args
                    .iter()
                    .map(|a| a.interpret(e, full_program))
                    .try_collect()?;
                o.execute(&args)?
            }
            ProgramNode::Rec(_) => {
                let args: Vec<Constant> = args
                    .iter()
                    .map(|a| a.interpret(e, full_program))
                    .try_collect()?;

                let new_e = &(&args).into();

                full_program.interpret(new_e, full_program)?
            }
            ProgramNode::If => {
                if args.get(0).unwrap().interpret(e, full_program)? == Constant::Bool(true) {
                    args.get(1).unwrap().interpret(e, full_program)?
                } else {
                    args.get(2).unwrap().interpret(e, full_program)?
                }
            }
        })
    }

    pub fn get_behavior(&self, tests: &[TestCase]) -> Vec<TestCase> {
        tests
            .iter()
            .filter_map(|t| {
                self.interpret(&t.into(), self).ok().map(|output| TestCase {
                    inputs: t.inputs.clone(),
                    output,
                })
            })
            .collect()
    }

    pub fn passes_test_case(&self, t: &TestCase) -> bool {
        self.interpret(&t.into(), self)
            .map_or(false, |output| output == t.output)
    }

    pub fn passes_all_test_cases(&self, v: &[TestCase]) -> bool {
        v.iter().all(|t| self.passes_test_case(t))
    }

    pub fn get_type(&self) -> T {
        match &self.node {
            ProgramNode::Constant(c) => c.clone().into(),
            ProgramNode::Variable(v) => v.ty.clone(),
            ProgramNode::Operation(o) => o.sig.output.clone(),
            ProgramNode::If => self.args.get(1).unwrap().get_type(),
            ProgramNode::Rec(t) => t.clone(),
        }
    }
}

impl<T: TypeSystemBounds> From<LinearProgram<T>> for Program<T> {
    fn from(LinearProgram { node, args }: LinearProgram<T>) -> Self {
        Program {
            node: node.into(),
            args: args.into_iter().map(|a| a.into()).collect(),
        }
    }
}

impl<T: TypeSystemBounds> SynthCostFunc for Program<T> {
    fn cost(&self) -> usize {
        match &self.node {
            ProgramNode::Variable(_) => 1,
            ProgramNode::Constant(_) => 2,
            ProgramNode::Operation(_) => 3 + self.args.iter().map(|a| a.cost()).sum::<usize>(),
            ProgramNode::If => 4 + self.args.iter().map(|a| a.cost()).sum::<usize>(),
            ProgramNode::Rec(_) => 5 + self.args.iter().map(|a| a.cost()).sum::<usize>(),
        }
    }
}

impl<T: TypeSystemBounds> Display for Program<T> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match &self.node {
            ProgramNode::Constant(c) => write!(f, "{c}"),
            ProgramNode::Variable(v) => write!(f, "{v}"),
            ProgramNode::Operation(o) => write!(
                f,
                "{o}({})",
                self.args.iter().map(ToString::to_string).join(",")
            ),
            ProgramNode::Rec(t) => write!(
                f,
                "rec<{t}>({})",
                self.args.iter().map(ToString::to_string).join(",")
            ),
            ProgramNode::If => write!(
                f,
                "if ({}) {{{}}} else {{{}}}",
                self.args.get(0).unwrap(),
                self.args.get(1).unwrap(),
                self.args.get(2).unwrap()
            ),
        }
    }
}
