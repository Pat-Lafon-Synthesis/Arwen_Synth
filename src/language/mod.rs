mod in_out_examples;
#[allow(clippy::module_inception)]
mod language;
mod sketches;
pub use in_out_examples::*;
pub use language::*;
pub use sketches::*;
