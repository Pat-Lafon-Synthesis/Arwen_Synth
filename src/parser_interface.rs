use std::ops::{Deref, DerefMut};

use program_synthesis_parser::lang::{
    ActualExp, ActualType, Example, Exp, Problem, SynthProblem, Type,
};

use crate::{
    language::{Constant, TestCase},
    types::{BaseType, Signature},
};

pub fn parse(src: String) -> MySynthProblem {
    let parser = program_synthesis_parser::spec::ProblemParser::new();
    parser.parse(&src).unwrap().into()
}

pub struct Examples {
    pub tests: Vec<TestCase>,
}

impl From<Vec<TestCase>> for Examples {
    fn from(tests: Vec<TestCase>) -> Self {
        Self { tests }
    }
}

impl From<Examples> for Vec<TestCase> {
    fn from(value: Examples) -> Self {
        value.tests
    }
}

impl Deref for Examples {
    type Target = Vec<TestCase>;

    fn deref(&self) -> &Self::Target {
        &self.tests
    }
}

impl DerefMut for Examples {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.tests
    }
}

impl From<&Problem> for Examples {
    fn from(value: &Problem) -> Self {
        match value {
            Problem::UIOEs(x) => Examples {
                tests: x.iter().map(Into::into).collect(),
            },
            Problem::UEquiv(..) | Problem::UPost(_) => unimplemented!(),
        }
    }
}

impl From<&Exp> for Constant {
    fn from(e: &Exp) -> Self {
        match &***e {
            ActualExp::CTor(name, exp) => match (name.as_ref(), &***exp) {
                ("True", ActualExp::Tuple(v)) if v.is_empty() => Constant::Bool(true),
                ("False", ActualExp::Tuple(v)) if v.is_empty() => Constant::Bool(false),
                ("O", ActualExp::Tuple(v)) if v.is_empty() => Constant::Int(0),
                ("S", ActualExp::CTor(..)) => {
                    if let Constant::Int(i) = (exp).into() {
                        Constant::Int(i + 1)
                    } else {
                        panic!()
                    }
                }
                ("Nil", ActualExp::Tuple(v)) if v.is_empty() => Constant::IntList(Vec::new()),
                ("Cons", ActualExp::Tuple(v)) if v.len() == 2 => {
                    if let (Constant::Int(n), Constant::IntList(mut v)) =
                        ((&v[0]).into(), (&v[1]).into())
                    {
                        v.push(n);
                        Constant::IntList(v)
                    } else {
                        panic!()
                    }
                }
                e => {
                    dbg!(&e);
                    unimplemented!()
                }
            },

            ActualExp::Var(_)
            | ActualExp::Wildcard
            | ActualExp::App(..)
            | ActualExp::Func(..)
            | ActualExp::Unctor(..)
            | ActualExp::Eq(..)
            | ActualExp::Match(..)
            | ActualExp::Fix(..)
            | ActualExp::Tuple(_)
            | ActualExp::Proj(..) => unimplemented!(),
        }
    }
}

impl From<&Example> for TestCase {
    fn from(Example { input, output }: &Example) -> Self {
        TestCase {
            inputs: input.iter().map(Into::into).collect(),
            output: output.into(),
        }
    }
}

pub struct MySynthProblem {
    pub sig: Signature<BaseType>,
    pub tests: Examples,
}

impl From<Type> for BaseType {
    fn from(value: Type) -> Self {
        match &**value {
            ActualType::Named(n) => match n.as_str() {
                "bool" => BaseType::Bool,
                "nat" => BaseType::Int,
                "list" => BaseType::IntList,
                _ => unimplemented!(),
            },
            ActualType::Arrow(..)
            | ActualType::Tuple(_)
            | ActualType::Mu(..)
            | ActualType::Variant(_) => unimplemented!(),
        }
    }
}

fn uncurry_arrow_type(t: Type) -> Vec<BaseType> {
    match &**t {
        ActualType::Arrow(t1, t2) => vec![
            uncurry_arrow_type(t1.clone()),
            uncurry_arrow_type(t2.clone()),
        ]
        .concat(),
        ActualType::Named(_) => vec![t.into()],
        ActualType::Tuple(_) | ActualType::Mu(..) | ActualType::Variant(_) => todo!(),
    }
}

impl From<Type> for Signature<BaseType> {
    fn from(value: Type) -> Self {
        match &**value {
            ActualType::Arrow(..) => {
                let mut typs = uncurry_arrow_type(value);
                let output = typs.pop().unwrap();
                Signature {
                    input: typs,
                    output,
                }
            }
            ActualType::Named(_)
            | ActualType::Tuple(_)
            | ActualType::Mu(..)
            | ActualType::Variant(_) => unimplemented!(),
        }
    }
}

impl From<SynthProblem> for MySynthProblem {
    fn from(
        SynthProblem {
            imports: _,
            decls: _,
            synth_type,
            spec,
        }: SynthProblem,
    ) -> Self {
        MySynthProblem {
            sig: synth_type.into(),
            tests: (&spec).into(),
        }
    }
}
