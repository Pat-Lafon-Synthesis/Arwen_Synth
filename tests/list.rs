use arwen_synth::synthesis;

use arwen_synth::language::Examples;
use arwen_synth::libraries::*;
use arwen_synth::parser_interface::parse;
use std::fs::File;
use std::io::Read;

macro_rules! make_test {
    ($test_name:tt, $($libs:tt)*) => {
        #[test]
        fn $test_name() {
            let mut file =
                File::open(format!("tests/benchmarks/{}.mls", stringify!($test_name))).unwrap();
            let mut buffer = String::new();
            file.read_to_string(&mut buffer).unwrap();
            let synth_problem = parse(buffer);

            let prog = synthesis(
                synth_problem.sig.into(),
                $($libs)*,
                Examples::new(synth_problem.tests.tests, Vec::new()),
                3,
            );
            insta::assert_display_snapshot!(stringify!($test_name), prog.unwrap());
        }
    };
}

make_test!(list_append, &list_library());
make_test!(list_compress, &list_library());
/* make_test!(list_concat, &list_library()); */
make_test!(list_drop, &list_nat_library());
make_test!(list_dropeven, &list_library());
make_test!(list_even_parity, &list_library());
make_test!(list_filter, &list_library());
make_test!(list_fold, &list_library());
make_test!(list_hd, &list_nat_library());
make_test!(list_inc, &list_library());
make_test!(list_last, &list_library());
make_test!(list_last2, &list_library());
make_test!(list_length, &list_nat_library());
make_test!(list_make, &list_library());
make_test!(list_map, &list_library());
make_test!(list_nth, &list_library());
make_test!(list_pairwise, &list_library());
make_test!(list_range, &list_library());
make_test!(list_rev_app, &list_library());
make_test!(list_rev_fold, &list_library());
make_test!(list_rev_snoc, &list_library());
make_test!(list_rev_tailcall, &list_library());
make_test!(list_snoc, &list_library());
make_test!(list_sort_sorted_insert, &list_library());
make_test!(list_sorted_insert, &list_library());
make_test!(list_stutter, &list_library());
make_test!(list_sum, &list_library());
make_test!(list_take, &list_library());
make_test!(list_tl, &list_library());
