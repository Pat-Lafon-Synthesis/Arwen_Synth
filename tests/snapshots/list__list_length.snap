---
source: tests/list.rs
expression: prog.unwrap()
---
if (is_nil(arg0)) {O()} else {inc(rec<int>(tail(arg0)))}
